README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# nomisr

[![License:
MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)
[![CRAN\_Status\_Badge](https://www.r-pkg.org/badges/version/nomisr)](https://cran.r-project.org/package=nomisr)
[![GitHub
tag](https://img.shields.io/github/tag/ropensci/nomisr.svg)](https://github.com/ropensci/nomisr)
[![](https://cranlogs.r-pkg.org/badges/grand-total/nomisr)](https://dgrtwo.shinyapps.io/cranview/)
[![Travis-CI Build
Status](https://travis-ci.org/ropensci/nomisr.svg?branch=master)](https://travis-ci.org/ropensci/nomisr)
[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/evanodell/nomisr?branch=master&svg=true)](https://ci.appveyor.com/project/evanodell/nomisr)
[![Coverage
Status](https://img.shields.io/codecov/c/github/ropensci/nomisr/master.svg)](https://codecov.io/github/ropensci/nomisr?branch=master)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1246845.svg)](https://doi.org/10.5281/zenodo.1246845)

`nomisr` is for accessing UK official statistics from the
[Nomis](https://www.nomisweb.co.uk/) database through R. Nomis contains
data from the Census, the Labour Force Survey, DWP benefit statistics
and other economic and demographic data, and is maintained on behalf of
the Office for National Statistics by the University of Durham.

The `nomisr` package provides functions to find what data is available,
the variables and query options for different datasets and a function
for downloading data. `nomisr` returns data in
[`tibble`](https://cran.r-project.org/package=tibble) format. Most of
the data available through `nomisr` is based around statistical
geographies, with a handful of exceptions.

The package is for demographers, economists, geographers, public health
researchers and any other researchers who are interested in geographic
factors. The package aims to aid reproducibility, reduce the need to
manually download area profiles, and allow easy linking of different
datasets covering the same geographic area.

## Installation

`nomisr` is available on CRAN:

``` r
install.packages("nomisr")
```

You can install the development version `nomisr` from github with:

``` r
# install.packages("devtools")
devtools::install_github("ropensci/nomisr")
```

## Using `nomisr`

`nomisr` contains functions to search for datasets, identify the query
options for different datasets and retrieve data from queries, all done
with [`tibbles`](http://tibble.tidyverse.org/), to take advantage of how
`tibble` manages list-columns. The use of metadata queries, rather than
simply downloading all available data, is useful to avoid overwhelming
the rate limits of the API. For full details on all available functions
and demonstrations of their use, please see the package
[vignette](https://docs.evanodell.com/nomisr/articles/introduction.html).

The example below gets the latest data on Jobseeker’s Allowance with
rates and proportions, on a national level, with all male claimants and
workforce.

``` r
 library(nomisr)
 jobseekers_search <- nomis_search(name = "*Jobseeker*")
 
 tibble::glimpse(jobseekers_search)
#> Observations: 17
#> Variables: 14
#> $ agencyid                             <chr> "NOMIS", "NOMIS", "NOMIS"...
#> $ id                                   <chr> "NM_1_1", "NM_4_1", "NM_8...
#> $ uri                                  <chr> "Nm-1d1", "Nm-4d1", "Nm-8...
#> $ version                              <dbl> 1, 1, 1, 1, 1, 1, 1, 1, 1...
#> $ annotations.annotation               <list> [<c("Current (being acti...
#> $ components.attribute                 <list> [<c("Mandatory", "Condit...
#> $ components.dimension                 <list> [<c("CL_1_1_GEOGRAPHY", ...
#> $ components.primarymeasure.conceptref <chr> "OBS_VALUE", "OBS_VALUE",...
#> $ components.timedimension.codelist    <chr> "CL_1_1_TIME", "CL_4_1_TI...
#> $ components.timedimension.conceptref  <chr> "TIME", "TIME", "TIME", "...
#> $ description.value                    <chr> "Records the number of pe...
#> $ description.lang                     <chr> "en", "en", NA, "en", "en...
#> $ name.value                           <chr> "Jobseeker's Allowance wi...
#> $ name.lang                            <chr> "en", "en", "en", "en", "...

 jobseekers_measures <- nomis_get_metadata("NM_1_1", "measures")
 
 tibble::glimpse(jobseekers_measures)
#> Observations: 4
#> Variables: 2
#> $ description <chr> "claimants", "workforce", "active", "residence"
#> $ value       <int> 20100, 20201, 20202, 20203
 
 jobseekers_geography <- nomis_get_metadata("NM_1_1", "geography", "TYPE")
 
 tail(jobseekers_geography)
#> # A tibble: 6 x 2
#>   description                                        value  
#>   <chr>                                              <chr>  
#> 1 government office regions tec / lec based          TYPE490
#> 2 government office regions (former inc. Merseyside) TYPE491
#> 3 standard statistical regions                       TYPE492
#> 4 pre-1996 local authority districts                 TYPE496
#> 5 pre-1996 counties / scottish regions               TYPE498
#> 6 countries                                          TYPE499
 
 jobseekers_sex <- nomis_get_metadata("NM_1_1", "sex", "TYPE")
 
 tibble::glimpse(jobseekers_sex)
#> Observations: 3
#> Variables: 2
#> $ description <chr> "Male", "Female", "Total"
#> $ value       <int> 5, 6, 7
 
 z <- nomis_get_data(id = "NM_1_1", time = "latest", geography = "TYPE499",
                     measures=c(20100, 20201), sex=5)
 
 tibble::glimpse(z)
#> Observations: 70
#> Variables: 34
#> $ DATE                <chr> "2018-04", "2018-04", "2018-04", "2018-04"...
#> $ DATE_NAME           <chr> "April 2018", "April 2018", "April 2018", ...
#> $ DATE_CODE           <chr> "2018-04", "2018-04", "2018-04", "2018-04"...
#> $ DATE_TYPE           <chr> "date", "date", "date", "date", "date", "d...
#> $ DATE_TYPECODE       <chr> "0", "0", "0", "0", "0", "0", "0", "0", "0...
#> $ DATE_SORTORDER      <chr> "0", "0", "0", "0", "0", "0", "0", "0", "0...
#> $ GEOGRAPHY           <chr> "2092957697", "2092957697", "2092957697", ...
#> $ GEOGRAPHY_NAME      <chr> "United Kingdom", "United Kingdom", "Unite...
#> $ GEOGRAPHY_CODE      <chr> "K02000001", "K02000001", "K02000001", "K0...
#> $ GEOGRAPHY_TYPE      <chr> "countries", "countries", "countries", "co...
#> $ GEOGRAPHY_TYPECODE  <chr> "499", "499", "499", "499", "499", "499", ...
#> $ GEOGRAPHY_SORTORDER <chr> "0", "0", "0", "0", "0", "0", "0", "0", "0...
#> $ SEX                 <chr> "5", "5", "5", "5", "5", "5", "5", "5", "5...
#> $ SEX_NAME            <chr> "Male", "Male", "Male", "Male", "Male", "M...
#> $ SEX_CODE            <chr> "5", "5", "5", "5", "5", "5", "5", "5", "5...
#> $ SEX_TYPE            <chr> "sex", "sex", "sex", "sex", "sex", "sex", ...
#> $ SEX_TYPECODE        <chr> "0", "0", "0", "0", "0", "0", "0", "0", "0...
#> $ SEX_SORTORDER       <chr> "0", "0", "0", "0", "0", "0", "0", "0", "0...
#> $ ITEM                <chr> "1", "1", "2", "2", "3", "3", "4", "4", "9...
#> $ ITEM_NAME           <chr> "Total claimants", "Total claimants", "Stu...
#> $ ITEM_CODE           <chr> "1", "1", "2", "2", "3", "3", "4", "4", "9...
#> $ ITEM_TYPE           <chr> "item", "item", "item", "item", "item", "i...
#> $ ITEM_TYPECODE       <chr> "0", "0", "0", "0", "0", "0", "0", "0", "0...
#> $ ITEM_SORTORDER      <chr> "0", "0", "1", "1", "2", "2", "3", "3", "4...
#> $ MEASURES            <chr> "20100", "20201", "20100", "20201", "20100...
#> $ MEASURES_NAME       <chr> "Persons claiming JSA", "Workplace-based e...
#> $ OBS_VALUE           <chr> "283937", "1.5", NA, NA, NA, NA, NA, NA, N...
#> $ OBS_STATUS          <chr> "A", "A", "Q", "Q", "Q", "Q", "Q", "Q", "Q...
#> $ OBS_STATUS_NAME     <chr> "Normal Value", "Normal Value", "These fig...
#> $ OBS_CONF            <chr> "F", "F", "F", "F", "F", "F", "F", "F", "F...
#> $ OBS_CONF_NAME       <chr> "Free (free for publication)", "Free (free...
#> $ URN                 <chr> "Nm-1d1d32292e0d2092957697d5d1d20100", "Nm...
#> $ RECORD_OFFSET       <chr> "0", "1", "2", "3", "4", "5", "6", "7", "8...
#> $ RECORD_COUNT        <chr> "70", "70", "70", "70", "70", "70", "70", ...
```

There is a lot of data available through Nomis, and there are some
limits to the amount of data that can be retrieved within a certain
period of time, although those are not published. For more details, see
the [full API documentation](https://www.nomisweb.co.uk/api/v01/help)
from Nomis. Full package documentation is available at
[docs.evanodell.com/nomisr](https://docs.evanodell.com/nomisr)

## Meta

Bug reports, suggestions, and code contributions are all welcome. Please
see [CONTRIBUTING.md](CONTRIBUTING.md) for details.

Please note that this project is released with a [Contributor Code of
Conduct](CONDUCT.md). By participating in this project you agree to
abide by its terms.

Please note that this project is not affiliated with the Office for
National Statistics or the University of Durham.

Get citation information for `nomisr` in R with `citation(package =
'nomisr')`

Odell E (2018). *nomisr: Access Nomis UK labour market data on with R*.
doi: 10.5281/zenodo.1246845 (URL:
<http://doi.org/10.5281/zenodo.1246845>), R package version 0.2.0,
\<URL: <https://github.com/ropensci/nomisr>\>.

A BibTeX entry for LaTeX users is

``` 
  @Manual{,
    title = {{nomisr}: Access Nomis UK Labour Market Data With R},
    author = {Evan Odell},
    year = {2018},
    note = {R package version 0.2.0},
    doi = {10.5281/zenodo.1157908},
    url = {https://github.com/ropensci/nomisr},
  }
```

License:
[MIT](LICENSE.md)

[![ropensci\_footer](https://ropensci.org/public_images/ropensci_footer.png)](https://ropensci.org)


NEWS.md
# nomisr 0.3.0

## New features and function changes

* New `nomis_codelist()` function, which returns the internal coding for 
different concepts used by the NOMIS API in a \code{tibble}, given a dataset 
ID and a concept name. 

* The `additional_queries` parameter in `nomis_get_data()` and 
`nomis_get_metadata()` has been deprecated and will eventually be removed. 
Please use the `...` parameter for queries including concepts not available 
through the default parameters.

* The `sex` parameter in `nomis_get_data()` will also work with datasets that 
use "gender" instead of "sex".

## Internal changes and bug fixes

* Uses `rsdmx` to parse metadata, fixing #7.

# nomisr 0.2.0

* Improved API key handling (#5)

* Increased test coverage

* Adding rOpenSci reviewers to DESCRIPTION file.


# nomisr 0.1.0

* Moved to rOpenSci github repository

* Added API key functionality, which is not required by the API but is 
useful for large requests.

* In interactive sessions, users are asked if they want to continue when 
calling more than 15 pages of data at a time.

# nomisr 0.0.2

* Introduction of additional parameters to the `nomis_get_data()` and 
`nomis_codes()` functions, improvements to documentation.

# nomisr 0.0.1

* 1st release. Rudimentary functions for retrieving information on available 
datasets and downloading datasets from nomis, with some limited parameters.



R/data_download.R


#' Retrieve Nomis datasets
#'
#' @description Retrieves specific datasets from Nomis, based on their ID. To
#' find dataset IDs, use \code{\link{nomis_data_info}}. Datasets are retrived
#' in csv format and parsed with the \code{read_csv} function from the
#' \code{readr} package into a tibble, with all columns parsed as character
#' columns.
#'
#' @description To find the code options for a given dataset, use
#' \code{\link{nomis_get_metadata}} for specific codes, and 
#' \code{\link{nomis_codelist}} for code values.
#'
#' @description This can be a slow process if querying significant amounts of
#' data. Guest users are limited to 25,000 rows per query, although
#' \code{nomisr} identifies queries that will return more than 25,000 rows,
#' sending individual queries and combining the results of those queries into
#' a single tibble. In interactive sessions, \code{nomisr} will warn you if 
#' guest users are requesting more than 350,000 rows of data, and if 
#' registered users are requesting more than 1,500,000 rows.
#'
#' @description Note the difference between the \code{time} and \code{date}
#' parameters.
#' The \code{time} and \code{date} parameters should not be used at the same
#' time. If they are, the function will retrieve data based on the the
#' \code{date} parameter. If given more than one query, \code{time} will
#' return all data available between those queries, inclusively, while
#' \code{date} will only return data for the exact queries specified. So
#' \code{time = c("first", "latest")} will return all data, while
#' \code{date = c("first", "latest")} will return only the first and latest
#' data published.
#'
#' @param id The ID of the dataset to retrieve.
#'
#' @param time Parameter for selecting dates and date ranges. There are two
#' styles of values that can be used to query time.
#'
#' The first is one or more of \code{"latest"} (returns the latest available
#' data), \code{"previous"} (the date prior to \code{"latest"}),
#' \code{"prevyear"} (the date one year prior to \code{"latest"}) or
#' \code{"first"} (the oldest available data for the dataset).
#'
#' The second style is to use or a specific date or multiple dates, in the
#' style of the time variable codelist, which can be found using the
#' \code{\link{nomis_get_metadata}} function.
#'
#' Values for the \code{time} and \code{date} parameters should not be used
#' at the same time. If they are, the function will retrieve data based
#' on the the \code{date} parameter.
#'
#' Defaults to \code{NULL}.
#'
#' @param date Parameter for selecting specific dates. There are two styles
#' of values that can be used to query time.
#'
#' The first is one or more of \code{"latest"} (returns the latest available
#' data), \code{"previous"} (the date prior to \code{"latest"}),
#' \code{"prevyear"} (the date one year prior to \code{"latest"}) or
#' \code{"first"} (the oldest available data for the dataset).
#'
#' The second style is to use or a specific date or multiple dates, in the
#' style of the time variable codelist, which can be found using the
#' \code{\link{nomis_get_metadata}} function.
#'
#' Values for the \code{time} and \code{date} parameters should not be used at
#' the same time. If they are, the function will retrieve data based on the
#' the \code{date} parameter.
#'
#' Defaults to \code{NULL}.
#'
#' @param geography The code of the geographic area to return data for. If
#' \code{NULL}, returns data for all available geographic areas, subject to
#' other parameters. Defaults to \code{NULL}. In the rare instance that a
#' geographic variable does not exist, if not \code{NULL}, the function
#' will return an error.
#'
#' @param measures The code for the statistical measure(s) to include in the
#' data. Accepts a single string or number, or a list of strings or numbers.
#' If \code{NULL}, returns data for all available statistical measures subject
#' to other parameters. Defaults to \code{NULL}.
#'
#' @param sex The code for sexes/genders to include in the dataset. 
#' Accepts a string or number, or a vector of strings or numbers. 
#' \code{nomisr} automatically voids any queries for sex if it is not an 
#' available code in the requested dataset.
#'
#' There are two different codings used for sex, depending on the dataset. For
#' datasets using \code{"SEX"}, \code{7} will return results for
#' males and females, \code{6} only females and \code{5} only males.
#' Defaults to \code{NULL}, equivalent to \code{c(5,6,7)} for datasets where
#' sex is an option. For datasets using \code{"C_SEX"}, \code{0} will return
#' results for males and females, \code{1} only males and
#' \code{2} only females. Some datasets use \code{"GENDER"} with the same 
#' values as \code{"SEX"}, which works with both \code{sex} and 
#' \code{gender = <code>} as a dot parameter.
#'
#' @param additional_queries Any other additional queries to pass to the API.
#' See \url{https://www.nomisweb.co.uk/api/v01/help} for instructions on
#' query structure. Defaults to \code{NULL}. Deprecated in package 
#' versions greater than 0.2.0 and will eventually be removed in a
#' future version.
#'
#' @param exclude_missing If \code{TRUE}, excludes all missing values.
#' Defaults to \code{FALSE}.
#'
#' @param select A character vector of one or more variables to select,
#' excluding all others. \code{select} is not case sensitive.
#' 
#' @param ... Use to pass any other parameters to the API. Useful for passing 
#' concepts that are not available through the default parameters. Only accepts
#' concepts identified in \code{\link{nomis_get_metadata}} and concept values 
#' identified in \code{\link{nomis_codelist}}. Parameters can be quoted or
#' unquoted, and are not case sensitive. Each parameter should have a name and
#' a value. For example \code{cause_of_death = 10300} and 
#' \code{CAUSE_OF_DEATH = 10300} will return the same result when querying 
#' dataset "NM_161_1".
#'
#' @return A tibble containing the selected dataset.
#' By default, all tibble columns are parsed as characters.
#' @export
#' @seealso \code{\link{nomis_data_info}}
#' @seealso \code{\link{nomis_get_metadata}}
#' @seealso \code{\link{nomis_codelist}}
#' @seealso \code{\link{nomis_overview}}
#' @examples \donttest{
#'
#' # Return data for each country
#' jobseekers_country <- nomis_get_data(id="NM_1_1", time="latest",
#'                                      geography = "TYPE499",
#'                                      measures=c(20100, 20201), sex=5)
#'
#' tibble::glimpse(jobseekers_country)
#'
#' # Return data for Wigan
#' jobseekers_wigan <- nomis_get_data(id="NM_1_1", time="latest",
#'                                    geography = "1879048226",
#'                                    measures=c(20100, 20201), sex="5")
#'
#' tibble::glimpse(jobseekers_wigan)
#'
#' # annual population survey - regional - employment by occupation
#' emp_by_occupation <- nomis_get_data(id="NM_168_1", time="latest",
#'                                     geography = "2013265925", sex="0",
#'                                     select = c("geography_code",
#'                                     "C_OCCPUK11H_0_NAME", "obs_vAlUE"))
#'
#' tibble::glimpse(emp_by_occupation)
#' 
#' # Deaths in 2016 and 2015 by three specified causes, g
#' death <- nomis_get_data("NM_161_1", date = c("2016","2015"), 
#'                         geography = "TYPE480", 
#'                         cause_of_death=c(10300, 102088, 270))
#'                     
#' tibble::glimpse(death)
#' 
#' 
#' # All causes of death in London in 2016
#' london_death <- nomis_get_data("NM_1_1", date = c("2016"), 
#'                                geography = "2013265927", sex = 1, age = 0)              
#'
#' tibble::glimpse(london_death)
#' }

nomis_get_data <- function(id, time = NULL, date = NULL, geography = NULL,
                           sex = NULL, measures = NULL, 
                           additional_queries = NULL, exclude_missing = FALSE,
                           select = NULL, ...) {
  if (missing(id)) {
    stop("Dataset ID must be specified", call. = FALSE)
  }

  # check for use or time or data parameter
  if (is.null(date) == FALSE) {
    time_query <- paste0("&date=", paste0(date, collapse = ","))
  } else if (is.null(time) == FALSE) {
    time_query <- paste0("&time=", paste0(time, collapse = ","))
  } else {
    time_query <- ""
  }

  geography_query <- ifelse(is.null(geography) == FALSE,
    paste0(
      "&geography=",
      paste0(geography, collapse = ",")
    ),
    ""
  )

  measures_query <- ifelse(length(measures) > 0,
    paste0(
      "&measures=",
      paste0(measures, collapse = ",")
    ),
    ""
  )

  if (length(additional_queries) > 0) {
    additional_query <- additional_queries

    message("The `additional_query` parameter is
            deprecated, please use ... instead")
  } else {
    additional_query <- NULL
  }

  # Check for sex queries and return either sex or c_sex
  if (length(sex) > 0) {
    sex_lookup <- nomis_data_info(id)$components.dimension[[1]]$conceptref

    if ("C_SEX" %in% sex_lookup) {
      sex_query <- paste0("&c_sex=", paste0(sex, collapse = ","))
    } else if ("SEX" %in% sex_lookup) {
      sex_query <- paste0("&sex=", paste0(sex, collapse = ","))
    }  else if ("GENDER" %in% sex_lookup) {
      sex_query <- paste0("&gender=", paste0(sex, collapse = ","))
    } else {
      sex_query <- ""
    }
  } else {
    sex_query <- ""
  }

  exclude_query <- ifelse(exclude_missing == TRUE,
    "&ExcludeMissingValues=true",
    ""
  )

  select_query <- ifelse(length(select) > 0,
    paste0(
      "&select=",
      paste0(
        unique(c(toupper(select), "RECORD_COUNT")),
        collapse = ","
      )
    ),
    ""
  )

  dots <- rlang::list2(...) ## eval the dots

  x <- c()
  
  for (i in seq_along(dots)) { # retrieve the dots
    x[i] <- ifelse(length(dots[[i]]) > 0,
      paste0(
        "&", names(dots[i]), "=",
        paste0(dots[[i]], collapse = ",")
      ),
      ""
    )
  }

  dots_query <- paste0(x, collapse = "")

  if (!is.null(getOption("nomisr.API.key"))) {
    api_query <- paste0("&uid=", getOption("nomisr.API.key"))
    max_length <- 100000
  } else {
    api_query <- ""
    max_length <- 25000
  }

  query <- paste0(
    id, ".data.csv?", time_query, geography_query, sex_query, measures_query,
    exclude_query, select_query, api_query, additional_query, dots_query
  )

  first_df <- nomis_get_data_util(query)

  if (nrow(first_df) == 0) {
    stop("The API request did not return any results.
         Please check your parameters.", call. = FALSE)
  }

  if (as.numeric(first_df$RECORD_COUNT)[1] >= max_length) {
    # if amount available is over the limit of 15 total calls at a time
    # downloads the extra data and binds it all together in a tibble

    if (interactive() &&
      as.numeric(first_df$RECORD_COUNT)[1] >= (15 * max_length)) {
      # For more than 15 total requests at one time.
      message(
        "Warning: You are trying to acess more than ",
        paste0((15 * max_length)), " rows of data."
      )
      message("This may cause timeout and/or automatic rate limiting.")

      if (utils::menu(c("Yes", "No"),
        title = "Do you want to continue?"
      ) == 2) {
        stop(call. = FALSE)
      }
    }

    record_count <- first_df$RECORD_COUNT[1]

    seq_list <- seq(from = max_length, to = record_count, by = max_length)

    pages <- list()

    for (i in seq_along(seq_list)) {
      query2 <- paste0(
        query, "&recordOffset=",
        format(seq_list[i], scientific = FALSE)
      )
      # R can paste large numbers as scientific notation, which causes an error
      # format(seq_list[i], scientific = FALSE)) prevents that

      message("Retrieving additional pages ", i, " of ", length(seq_list))

      pages[[i]] <- nomis_get_data_util(query2)
    }

    df <- tibble::as_tibble(dplyr::bind_rows(first_df, pages))
  } else {
    df <- first_df
  }

  if (length(select) > 0 & !("RECORD_COUNT" %in% select)) {
    df$RECORD_COUNT <- NULL
  }

  df
}
